src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,714 @@
+        assets = convertToAssets(shares);
+    }
+
+    /// @notice Request to redeem liquidity provider shares for liquidity assets
+    /// @param shares The amount of shares the redeemer wants to burn for liquidity assets
+    /// @return requestId The index of the withdrawal request
+    /// @return assets The amount of liquidity assets that will be claimable by the redeemer
+    function requestRedeem(uint256 shares) external returns (uint256 requestId, uint256 assets) {
+        // Calculate the amount of assets to transfer to the redeemer
+        assets = convertToAssets(shares);
+
+        requestId = nextWithdrawalIndex;
+        uint120 queued = SafeCast.toUint120(withdrawsQueued + assets);
+        uint40 claimTimestamp = uint40(block.timestamp + claimDelay);
+
+        // Store the next withdrawal request
+        nextWithdrawalIndex = SafeCast.toUint16(requestId + 1);
+        // Store the updated queued amount which reserves liquidity assets (WETH) in the withdrawal queue
+        withdrawsQueued = queued;
+        // Store requests
+        withdrawalRequests[requestId] = WithdrawalRequest({
+            withdrawer: msg.sender,
+            claimed: false,
+            claimTimestamp: claimTimestamp,
+            assets: SafeCast.toUint120(assets),
+            queued: queued
+        });
+
+        // burn redeemer's shares
+        _burn(msg.sender, shares);
+
+        // Remove the redeemed assets from the last available assets
+        lastAvailableAssets -= SafeCast.toInt128(SafeCast.toInt256(assets));
+
+        emit RedeemRequested(msg.sender, requestId, assets, queued, claimTimestamp);
+    }
+
+    /// @notice Claim liquidity assets from a previous withdrawal request after the claim delay has passed
+    /// @param requestId The index of the withdrawal request
+    /// @return assets The amount of liquidity assets that were transferred to the redeemer
+    function claimRedeem(uint256 requestId) external returns (uint256 assets) {
+        // Load the structs from storage into memory
+        WithdrawalRequest memory request = withdrawalRequests[requestId];
+
+        require(request.claimTimestamp <= block.timestamp, "Claim delay not met");
+        // Is there enough liquidity to claim this request?
+        require(
+            request.queued <= withdrawsClaimed + IERC20(liquidityAsset).balanceOf(address(this)),
+            "Queue pending liquidity"
+        );
+        require(request.withdrawer == msg.sender, "Not requester");
+        require(request.claimed == false, "Already claimed");
+
+        assets = request.assets;
+
+        // Store the request as claimed
+        withdrawalRequests[requestId].claimed = true;
+        // Store the updated claimed amount
+        withdrawsClaimed += SafeCast.toUint120(assets);
+
+        // transfer the liquidity asset to the withdrawer
+        IERC20(liquidityAsset).transfer(msg.sender, assets);
```

Comment:
Do you have a test that can replicate this scenario?

In theory, if this first require passes for this request or any higher requests in the queue, then there should be enough liquidity.
```Solidity
require(
    request.queued <= withdrawsClaimed + IERC20(liquidityAsset).balanceOf(address(this)),
    "Queue pending liquidity"
);
```

This check is basically saying can the redeemer claim their position in the withdrawal queue with the available WETH liquidity. `request.queued` is the queue height at the time of their request and includes their `request.assets` amount. `withdrawsClaimed` is everyone who has already claimed. So `request.queued - withdrawsClaimed` should give what is outstanding to their position in the withdrawal queue.
If there are no other outstanding requests, then `request.queued - withdrawsClaimed == request.assets`.
If there are other requests before them in the queue, then `request.queued - withdrawsClaimed > request.assets`.

Let's look at some examples. There are the following unclaimed requests in the queue:
1. 2 WETH
2. 3 WETH
3. 6 WETH

If there was 1 WETH in the ARM, then no requests can be claimed
If there was 2.5 WETH in the ARM, then only the first request can be claimed
If there was 7 WETH in the ARM, then the first and second requests can be claimed
If there was 12 WETH in the ARM, then all three can be claimed

That's simple enough, but let's say we have this scenario
1. 2 WETH claimed, request.queue = 2
2. 3 WETH unclaimed, request.queue = 5
3. 6 WETH unclaimed, request.queue = 11
withdrawsQueued = 11
withdrawsClaimed = 2

So the second request can only be claimed when 
request.queued - withdrawsClaimed <= the ARM's WETH balance
So there needs to be 5 - 2 = 3 or more WETH in the ARM for the second request to be claimed.

What if a higher request has already been claimed? Let's say we have this
1. 2 WETH claimed, request.queue = 2
2. 3 WETH unclaimed, request.queue = 5
3. 6 WETH claimed, request.queue = 11
withdrawsQueued = 11
withdrawsClaimed = 8

So for the second request to be claimed, there needs to be request.queued - withdrawsClaimed = 5 - 8 = -3 WETH
But, in order for the third request to be claimed, there must have been request.queued - withdrawsClaimed = 11 - 2 = 9 WETH available at the time.
And since we are reserving that WETH for the withdrawal queue, there should still be 9 - 6 = 3 WETH still available. That's enough for the second request to be claimed.

The key for the withdrawal queue to work is not using the WETH once it becomes available for the withdrawal queue.

Reply:
Thanks a lot for taking the time to explain this! 

-Indeed, only checking available liquidity cannot replace this check.

- My first question about the fact that it cannot revert properly was whether a user could swap out all the WETH in the ARM. I've tried to write a test for it, and it is not possible to take WETH that are reserved for queue when swapping, my test failed. 

So all good, thanks a lot for the explanation 🙏